pkg/identity/identity.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/yahoo/k8s-athenz-syncer/pkg/log"
+)
+
+var (
+	gracePeriod = 5 * time.Minute
+)
+
+// PrivateKeyProvider returns a signing key on demand.
+type PrivateKeyProvider func() (*crypto.SigningKey, error)
+
+// tokenProvider provides tokens.
+type tokenProvider interface {
+	Token() (string, error)
+}
+
+// TokenProvider implements tokenProvider provides nTokens
+type TokenProvider struct {
+	client  *zms.ZMSClient
+	header  string
+	domain  string
+	service string
+	expiry  time.Duration
+	ks      PrivateKeyProvider
+	l       sync.RWMutex
+	current string
+	expire  time.Time
+}
+
+// Config is the token provider configuration.
+type Config struct {
+	Client             *zms.ZMSClient
+	Header             string
+	Domain             string             // Athenz domain
+	Service            string             // Athenz service
+	PrivateKeyProvider PrivateKeyProvider // source for private keys
+	TokenExpiry        time.Duration      // token expire time
+}
+
+// NewTokenProvider returns a token for the supplied configuration.
+func NewTokenProvider(config Config, stopCh <-chan struct{}) (*TokenProvider, error) {
+	if config.TokenExpiry == 0 {
+		config.TokenExpiry = time.Hour
+	}
+
+	tp := &TokenProvider{
+		client:  config.Client,
+		header:  config.Header,
+		domain:  config.Domain,
+		service: config.Service,
+		expiry:  config.TokenExpiry,
+		ks:      config.PrivateKeyProvider,
+	}
+	if _, err := tp.Token(); err != nil {
+		return nil, errors.Wrap(err, "mint token")
+	}
+	go tp.refreshLoop(time.Duration(config.TokenExpiry/3), stopCh)
+	return tp, nil
+}
+
+// updateToken - creates new nToken after there was no token or the existing token has expired
+func (tp *TokenProvider) updateToken() error {
+	key, err := tp.ks()
+	if err != nil {
+		return err
+	}
+	pem, err := key.PEMBytes()
+	if err != nil {
+		return err
+	}
+	b, err := zmssvctoken.NewTokenBuilder(tp.domain, tp.service, pem, key.Version)
+	if err != nil {
+		return err
+	}
+	b.SetExpiration(tp.expiry)
+	tok, err := b.Token().Value()
+	if err != nil {
+		return err
+	}
+	tp.current = tok
+	tp.expire = time.Now().Add(tp.expiry).Add(-1 * gracePeriod)
+	tp.client.AddCredentials(tp.header, tp.current)
+	log.Info("New nToken generated and added to zmsClient credentials")
+	return nil
+}
+
+// Token implements the tokenProvider interface
+func (tp *TokenProvider) Token() (string, error) {
+	tp.l.Lock()
+	defer tp.l.Unlock()
+	now := time.Now().Add(gracePeriod)
+	if tp.expire.Before(now) {
+		log.Info("Current NToken expired, getting ready to refresh")
+		if err := tp.updateToken(); err != nil {
+			return "", err
+		}
+	}
+	return tp.current, nil
+}
+
+// refreshLoop go subroutine that checks if the current token is expired and updates the n token
+func (tp *TokenProvider) refreshLoop(interval time.Duration, stopCh <-chan struct{}) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if _, err := tp.Token(); err != nil {
```

Comment:
should we call updateToken() instead of Token() here? Token() might still not refresh until the expiry?

Reply:
The refreshLoop() function runs more frequently than token expire duration (refreshLoop runs every (token.expiry duration)/3). If token is not expired at the time refreshLoop is run, token will not get updated. The check whether a token is expired is in Token() and UpdateToken() has pure logic for creating a new token (no expiration check)

Reply:
Would there be a time when the token has expired, until one of the refresh loop catches it? Would requests to ZMS fail during expired time?